GreedyParameterPointSampler.h

```diff
@@ -0,0 +1,734 @@
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: This class greedily selects parameter points
+//              for the construction of a ROM database. The implemented
+//              greedy algorithm is derived from algorithm 2 of Choi et. al's
+//              paper "Gradient-based constrained optimization using a database
+//              of linear reduced-order models": https://arxiv.org/abs/1506.07849
+//
+//              The greedy algorithm workflow is as follows:
+//              1. Construct the GreedyParameterPointSampler by giving it a
+//                 domain of parameter points.
+//              2. Request a parameter point to sample.
+//              3. Request a parameter point to compute an error error indicator for.
+//              4. Give the computed error indicator to the GreedyParameterPointSampler.
+//              5. Repeat steps 3 and 4 until the GreedyParameterPointSampler
+//                 no longer requires more error indicators to be computed.
+//              6. Repeat steps 2 to 5 until the GreedyParameterPointSampler
+//                 no longer requires more parameter points to be sampled.
+//              7. The ROM database is now complete, meeting the error tolerance
+//                 for all parameter points within the domain.
+
+#ifndef included_GreedyParameterPointSampler_h
+#define included_GreedyParameterPointSampler_h
+
+#include "Vector.h"
+#include <string>
+#include <vector>
+#include <set>
+#include <random>
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+
+namespace CAROM {
+
+class Vector;
+
+struct GreedyErrorIndicatorPoint {
+    std::shared_ptr<Vector> point;
+    std::shared_ptr<Vector> localROM;
+};
+
+/**
+ * Class GreedyParameterPointSampler is a class defining the interface of a
+ *       greedy algorithm that given a domain of parameter points, iteratively
+ *       returns the next best parameter point to sample in order to create
+ *       a ROM database efficiently.
+ */
+class GreedyParameterPointSampler
+{
+public:
+    /**
+     * @brief Constructor.
+     *
+     * @param[in] parameter_points A vector of CAROM::Vectors containing
+     *                             the different parameter points.
+     * @param[in] check_local_rom Compute local ROM error indicator each iteration.
+     * @param[in] relative_error_tolerance The relative error tolerance value
+     *                                     for which to end the algorithm.
+     * @param[in] alpha A alpha constant to increase greedy algorithm by each
```

Comment:
Would it be more accurate to say "A constant factor to increase the error indicator tolerance"?